onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -512,208 +512,348 @@ Status EmbedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_l
     }
     Node& layer_norm_add_node = *graph.GetNode(edges[0]->GetNode().Index());
 
-    // Trace back to find the Gather for segment embedding.
-    std::vector<graph_utils::EdgeEndToMatch> segment_embedding_path{
-        {0, 1, "Gather", {1, 11}, kOnnxDomain}};
-    if (!graph_utils::FindPath(layer_norm_add_node, true, segment_embedding_path, edges, logger)) {
-      continue;
-    }
-    Node& segment_gather_node = *graph.GetNode(edges[0]->GetNode().Index());
-    if (!optimizer_utils::CheckOutputEdges(graph, segment_gather_node, 1)) {
-      continue;
-    }
-    // The first input of segment_gather_node must be 2d.
-    NodeArg* segment_embedding = segment_gather_node.MutableInputDefs()[0];
-    auto sg_shape = segment_embedding->Shape();
-    if (sg_shape == nullptr || sg_shape->dim_size() != 2 ||
-        !utils::HasDimValue(sg_shape->dim()[1]) ||
-        sg_shape->dim()[1].dim_value() <= 0) {
-      continue;
+    // Check if it's distil-bert
+    int add_count = 0;
+    int gather_count = 0;
+    for (auto it = layer_norm_add_node.InputNodesBegin(); it != layer_norm_add_node.InputNodesEnd(); ++it) {
+      if ((*it).OpType().compare("Add") == 0) {
+        add_count++;
+      } else if ((*it).OpType().compare("Gather") == 0) {
+        gather_count++;
+      }
     }
-    auto hidden_size = sg_shape->dim()[1].dim_value();
 
-    // Trace back to find Gather --> Add --> LayerNormalization
-    std::vector<graph_utils::EdgeEndToMatch> word_embedding_path{
-        {0, 0, "Add", {7}, kOnnxDomain},
-        {0, 0, "Gather", {1, 11}, kOnnxDomain}};
-    if (!graph_utils::FindPath(layer_norm_add_node, true, word_embedding_path, edges, logger)) {
-      continue;
-    }
-    Node& add_node = *graph.GetNode(edges[0]->GetNode().Index());
-    Node& word_gather_node = *graph.GetNode(edges[1]->GetNode().Index());
-    if (!optimizer_utils::CheckOutputEdges(graph, add_node, 1) ||
-        !optimizer_utils::CheckOutputEdges(graph, word_gather_node, 1)) {
-      continue;
-    }
-    // The first input of word_gather_node must be 2d.
-    NodeArg* word_embedding = word_gather_node.MutableInputDefs()[0];
-    auto wg_shape = word_embedding->Shape();
-    if (wg_shape == nullptr || wg_shape->dim_size() != 2 ||
-        !utils::HasDimValue(wg_shape->dim()[1]) ||
-        wg_shape->dim()[1].dim_value() != hidden_size) {
-      DEBUG_LOG("Word embedding shape not expected.");
-      continue;
-    }
+    if (add_count == 0 && gather_count == 2) {
```

Comment:
Instead of checking node count, why not check that the add node has two input nodes, and both are Gather directly.